src/runtime/transaction_execution.zig

```diff
@@ -0,0 +1,586 @@
+        err: TransactionError,
+    };
+}
+
+pub const TransactionFees = struct {
+    transaction_fee: u64,
+    prioritization_fee: u64,
+};
+
+pub const TransactionRollbacks = union(enum(u8)) {
+    fee_payer_only: CopiedAccount,
+    same_nonce_and_fee_payer: CopiedAccount,
+    separate_nonce_and_fee_payer: struct { CopiedAccount, CopiedAccount },
+};
+
+pub const CopiedAccount = struct {
+    pubkey: Pubkey,
+    account: AccountSharedData,
+};
+
+pub const ExecutedTransaction = struct {
+    err: ?TransactionError,
+    log_collector: ?LogCollector,
+    instruction_trace: ?InstructionTrace,
+    return_data: ?TransactionReturnData,
+    compute_meter: u64,
+    accounts_data_len_delta: i64,
+
+    pub fn deinit(self: ExecutedTransaction, allocator: std.mem.Allocator) void {
+        if (self.log_collector) |lc| lc.deinit(allocator);
+        if (self.return_data) |data| allocator.free(data);
+    }
+};
+
+pub const ProcessedTransaction = union(enum(u8)) {
+    fees_only: struct {
+        err: TransactionError,
+        fees: TransactionFees,
+        rollbacks: TransactionRollbacks,
+    },
+    executed: struct {
+        loaded_accounts: LoadedTransactionAccounts,
+        executed_transaction: ExecutedTransaction,
+    },
+
+    pub fn deinit(self: ProcessedTransaction, allocator: std.mem.Allocator) void {
+        switch (self) {
+            .executed => |executed| {
+                executed.loaded_accounts.deinit(allocator);
+                executed.executed_transaction.deinit(allocator);
+            },
+            else => {},
+        }
+    }
+};
+
+/// [agave] https://github.com/firedancer-io/agave/blob/403d23b809fc513e2c4b433125c127cf172281a2/svm/src/transaction_processor.rs#L323-L324
+pub fn loadAndExecuteTransactions(
+    allocator: std.mem.Allocator,
+    transactions: []const RuntimeTransaction,
+    batch_account_cache: *BatchAccountCache,
+    environment: *const TransactionExecutionEnvironment,
+    config: *const TransactionExecutionConfig,
+) error{OutOfMemory}![]TransactionResult(ProcessedTransaction) {
+    const transaction_results = try allocator.alloc(
+        TransactionResult(ProcessedTransaction),
+        transactions.len,
+    );
+    for (transaction_results, transactions) |result, *transaction| {
+        result.* = try loadAndExecuteTransaction(
+            allocator,
+            transaction,
+            batch_account_cache,
+            environment,
+            config,
+        );
+    }
+    return transaction_results;
+}
+
+/// [agave] https://github.com/firedancer-io/agave/blob/403d23b809fc513e2c4b433125c127cf172281a2/svm/src/transaction_processor.rs#L323-L324
+pub fn loadAndExecuteTransaction(
+    allocator: std.mem.Allocator,
+    transaction: *const RuntimeTransaction,
+    batch_account_cache: *BatchAccountCache,
+    environment: *const TransactionExecutionEnvironment,
+    config: *const TransactionExecutionConfig,
+) error{OutOfMemory}!TransactionResult(ProcessedTransaction) {
```

Comment:
This return value is a bit confusing because you have an error case in the first layer, and then in the next layer, fees only transactions always have an error, and executed transactions can also have errors. What is the difference between TransactionResult.err and fees_only? Can you document the return value here? Alternatively, what do you think about adding `err` to ProcessedTransaction (maybe with a more specific name) and removing the TransactionResult layer? Then the type can be documented centrally and speak for itself.

Reply:
The first layer is just for OutOfMemory, in which case this is our problem, not the transaction's problem. 

But I see what you mean with the Result ProcessedTransaction stuff. I think this just mirrors how Agave's error handling is structured, might require another look